## **Open Payments 2020 General Payments Data : Physician Information for the Doctors Table**

In [ ]:
# packages to import
import numpy as np
import pandas as pd

The Open Payments data full csv file with 91 columns and millions of rows:
* 2020 (3.7GB) https://download.cms.gov/openpayments/PGYR20_P012023/OP_DTL_GNRL_PGYR2020_P01202023.csv



In [ ]:
from google.colab import drive
drive.mount('/content/gdrive/', force_remount=True)

Mounted at /content/gdrive/


In [ ]:
# My Team3Folder was in my top level of my Google drive. If yours isn't, then you just need to modify it to add the Folder hierarchy
%cd /content/gdrive/MyDrive/Team3Folder/Data/Raw

/content/gdrive/MyDrive/Team3Folder/Raw Data


In [ ]:
# Read in columns needed for our data frame
df=pd.read_csv("OP_DTL_GNRL_PGYR2020_P01202023.csv", usecols=['Covered_Recipient_Profile_ID',
                  'Covered_Recipient_NPI',
                  'Covered_Recipient_First_Name',
                  'Covered_Recipient_Last_Name',
                  'Recipient_City',
                  'Recipient_State',
                  'Covered_Recipient_Type',
                  'Covered_Recipient_Primary_Type_1',
                  'Covered_Recipient_Specialty_1'])
df.shape

(5823162, 9)

In [ ]:
# Retrieve the data for Physicians only then drop the column Covered_Recipient_Type
df_doctors=df[df["Covered_Recipient_Type"]=="Covered Recipient Physician"]
df_doctors.shape

(5790756, 9)

In [ ]:
# Remove the duplicate rows if the Physicians have the same identifying information of Profile Id, NPI, First Name and Last Name
df_unique_doctors=df_doctors.drop_duplicates( subset=['Covered_Recipient_Profile_ID','Covered_Recipient_NPI','Covered_Recipient_First_Name','Covered_Recipient_Last_Name'],keep=False, inplace=False)
print(df_unique_doctors.shape)
df_unique_doctors.head(10)

(199242, 9)


Covered_Recipient_Type  Covered_Recipient_Profile_ID  \
0     Covered Recipient Physician                      557946.0   
1     Covered Recipient Physician                      276936.0   
2     Covered Recipient Physician                     1275463.0   
3     Covered Recipient Physician                      268352.0   
4     Covered Recipient Physician                      904225.0   
10    Covered Recipient Physician                     1044238.0   
1746  Covered Recipient Physician                      301917.0   
1750  Covered Recipient Physician                      461633.0   
1769  Covered Recipient Physician                     1235573.0   
1807  Covered Recipient Physician                       62540.0   

      Covered_Recipient_NPI Covered_Recipient_First_Name  \
0              1.841302e+09                        Vikas   
1              1.093896e+09                      Matthew   
2              1.154534e+09                      Stephen   
3              1.902872e+09                        Leroy   
4              1.003949e+09                      Michael   
10             1.912015e+09                       Joseph   
1746           1.669499e+09                     Kathleen   
1750           1.447482e+09                      Brianna   
1769           1.639513e+09                        YUSRA   
1807           1.376603e+09                         EMMA   

     Covered_Recipient_Last_Name Recipient_City Recipient_State  \
0                          Pilly  Williamsville              NY   
1                           Hall       Syracuse              NY   
2                       Campbell       Hinsdale              IL   
3                        McCarty          Edina              MN   
4                          Latza     Plainfield              NJ   
10                         Rosen        Lebanon              NH   
1746                        Cain         Topeka              KS   
1750                        Rhue        Tamarac              FL   
1769                    SIDDIQUI      GALVESTON              TX   
1807                      LOUCKS      GALVESTON              TX   

     Covered_Recipient_Primary_Type_1  \
0                      Medical Doctor   
1                 Doctor of Dentistry   
2                 Doctor of Dentistry   
3                      Medical Doctor   
4                        Chiropractor   
10                     Medical Doctor   
1746                   Medical Doctor   
1750              Doctor of Optometry   
1769                   Medical Doctor   
1807                   Medical Doctor   

                          Covered_Recipient_Specialty_1  
0     Allopathic & Osteopathic Physicians|Pain Medic...  
1                              Dental Providers|Dentist  
2                              Dental Providers|Dentist  
3     Allopathic & Osteopathic Physicians|Orthopaedi...  
4                   Chiropractic Providers|Chiropractor  
10    Allopathic & Osteopathic Physicians|Plastic Su...  
1746     Allopathic & Osteopathic Physicians|Pediatrics  
1750      Eye and Vision Services Providers|Optometrist  
1769  Allopathic & Osteopathic Physicians|Ophthalmology  
1807  Allopathic & Osteopathic Physicians|Ophthalmology

In [ ]:
#Makes sense to  rename the rest of the columns to simpler names
dict={
    "Covered_Recipient_Profile_ID":"ProfileID",
    "Covered_Recipient_NPI":"NPI",
    "Covered_Recipient_First_Name":"FirstName",
    "Covered_Recipient_Last_Name":"LastName",
    "Recipient_City":"City",
    "Recipient_State":"State",
    "Covered_Recipient_Type":"Type",
    "Covered_Recipient_Primary_Type_1":"Primary",
    "Covered_Recipient_Specialty_1":"Specialty"}
df_unique_doctors.rename(columns=dict,inplace=True)
df_unique_doctors.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 199242 entries, 0 to 5823155
Data columns (total 9 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   Type       199242 non-null  object 
 1   ProfileID  199242 non-null  float64
 2   NPI        198614 non-null  float64
 3   FirstName  199240 non-null  object 
 4   LastName   199233 non-null  object 
 5   City       199242 non-null  object 
 6   State      199220 non-null  object 
 7   Primary    199242 non-null  object 
 8   Specialty  199191 non-null  object 
dtypes: float64(2), object(7)
memory usage: 15.2+ MB


<ipython-input-7-b1e7149db9ce>:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_unique_doctors.rename(columns=dict,inplace=True)


In [ ]:
#Now that we have reduced the size, Lets look at the columns
df_unique_doctors.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 199242 entries, 0 to 5823155
Data columns (total 9 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   Type       199242 non-null  object 
 1   ProfileID  199242 non-null  float64
 2   NPI        198614 non-null  float64
 3   FirstName  199240 non-null  object 
 4   LastName   199233 non-null  object 
 5   City       199242 non-null  object 
 6   State      199220 non-null  object 
 7   Primary    199242 non-null  object 
 8   Specialty  199191 non-null  object 
dtypes: float64(2), object(7)
memory usage: 15.2+ MB


In [ ]:
# Lets find how how many nulls are in each column
df_unique_doctors.isna().sum()

Type           0
ProfileID      0
NPI          628
FirstName      2
LastName       9
City           0
State         22
Primary        0
Specialty     51
dtype: int64

199,242 doctors is a much more reasonable list to work with. If we drop the rows of doctors with nulls for the NPI value, we can then use the NPI value as the Primary Key in the doctors table since it should be a column in the non-OpenPayments data

In [ ]:
# delete all 628 rows with column 'NPI' being Null
df_unique_doctors = df_unique_doctors.dropna(subset=['NPI'])
df_unique_doctors.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 198614 entries, 0 to 5823155
Data columns (total 9 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   Type       198614 non-null  object 
 1   ProfileID  198614 non-null  float64
 2   NPI        198614 non-null  float64
 3   FirstName  198612 non-null  object 
 4   LastName   198605 non-null  object 
 5   City       198614 non-null  object 
 6   State      198593 non-null  object 
 7   Primary    198614 non-null  object 
 8   Specialty  198563 non-null  object 
dtypes: float64(2), object(7)
memory usage: 15.2+ MB


In [ ]:
# Lets find how how many nulls are left in each column
df_unique_doctors.isna().sum()

In [ ]:
# ProfileID and NPI columns are the only numeric data types, lets take a look
df_unique_doctors[['ProfileID', 'NPI']].head

<bound method NDFrame.head of          ProfileID           NPI
0         557946.0  1.841302e+09
1         276936.0  1.093896e+09
2        1275463.0  1.154534e+09
3         268352.0  1.902872e+09
4         904225.0  1.003949e+09
...            ...           ...
5823107   367896.0  1.164450e+09
5823113  1373001.0  1.063897e+09
5823120    30597.0  1.689604e+09
5823126  1123751.0  1.700809e+09
5823155  1391607.0  1.760625e+09

[198614 rows x 2 columns]>

In [ ]:
# It looks like they are composed of integer values represented as decimals. Lets convert these columns to ints.
df_unique_doctors[['ProfileID', 'NPI']]=df_unique_doctors[['ProfileID', 'NPI']].astype(int)
df_unique_doctors.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 198614 entries, 0 to 5823155
Data columns (total 9 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   Type       198614 non-null  object
 1   ProfileID  198614 non-null  int64 
 2   NPI        198614 non-null  int64 
 3   FirstName  198612 non-null  object
 4   LastName   198605 non-null  object
 5   City       198614 non-null  object
 6   State      198593 non-null  object
 7   Primary    198614 non-null  object
 8   Specialty  198563 non-null  object
dtypes: int64(2), object(7)
memory usage: 15.2+ MB


In [ ]:
df_unique_doctors[['ProfileID', 'NPI']].head(10)

ProfileID         NPI
0        557946  1841302403
1        276936  1093895625
2       1275463  1154534287
3        268352  1902871882
4        904225  1003949256
10      1044238  1912015215
1746     301917  1669498663
1750     461633  1447482278
1769    1235573  1639513211
1807      62540  1376602524

In [ ]:
#Now lets reset the index of the df_unique_doctors data frame to  the NPI column
df_unique_doctors.set_index('NPI', inplace=True)
df_unique_doctors.head()

Type  ProfileID FirstName  LastName  \
NPI                                                                      
1841302403  Covered Recipient Physician     557946     Vikas     Pilly   
1093895625  Covered Recipient Physician     276936   Matthew      Hall   
1154534287  Covered Recipient Physician    1275463   Stephen  Campbell   
1902871882  Covered Recipient Physician     268352     Leroy   McCarty   
1003949256  Covered Recipient Physician     904225   Michael     Latza   

                     City State              Primary  \
NPI                                                    
1841302403  Williamsville    NY       Medical Doctor   
1093895625       Syracuse    NY  Doctor of Dentistry   
1154534287       Hinsdale    IL  Doctor of Dentistry   
1902871882          Edina    MN       Medical Doctor   
1003949256     Plainfield    NJ         Chiropractor   

                                                    Specialty  
NPI                                                            
1841302403  Allopathic & Osteopathic Physicians|Pain Medic...  
1093895625                           Dental Providers|Dentist  
1154534287                           Dental Providers|Dentist  
1902871882  Allopathic & Osteopathic Physicians|Orthopaedi...  
1003949256                Chiropractic Providers|Chiropractor

In [ ]:
# I notices the FirstName, LastName and City columns have mixed case. To meake it simpler, and based on the dashboard, lets make them consistent
df_unique_doctors['FirstName']=df_unique_doctors['FirstName'].str.capitalize()
df_unique_doctors['LastName']=df_unique_doctors['LastName'].str.capitalize()
df_unique_doctors['City']=df_unique_doctors['City'].str.capitalize()
df_unique_doctors

Type  ProfileID FirstName  LastName  \
NPI                                                                      
1841302403  Covered Recipient Physician     557946     Vikas     Pilly   
1093895625  Covered Recipient Physician     276936   Matthew      Hall   
1154534287  Covered Recipient Physician    1275463   Stephen  Campbell   
1902871882  Covered Recipient Physician     268352     Leroy   Mccarty   
1003949256  Covered Recipient Physician     904225   Michael     Latza   
...                                 ...        ...       ...       ...   
1164449849  Covered Recipient Physician     367896     Bruce    Runyon   
1063896868  Covered Recipient Physician    1373001   Timothy      Ryan   
1689604191  Covered Recipient Physician      30597    Thomas   Schiano   
1700808862  Covered Recipient Physician    1123751   Pratima    Sharma   
1760625487  Covered Recipient Physician    1391607     Ethan  Weinberg   

                     City State              Primary  \
NPI                                                    
1841302403  Williamsville    NY       Medical Doctor   
1093895625       Syracuse    NY  Doctor of Dentistry   
1154534287       Hinsdale    IL  Doctor of Dentistry   
1902871882          Edina    MN       Medical Doctor   
1003949256     Plainfield    NJ         Chiropractor   
...                   ...   ...                  ...   
1164449849     Loma linda    CA       Medical Doctor   
1063896868        Jackson    MS       Medical Doctor   
1689604191       New york    NY       Medical Doctor   
1700808862      Ann arbor    MI       Medical Doctor   
1760625487   Philadelphia    PA       Medical Doctor   

                                                    Specialty  
NPI                                                            
1841302403  Allopathic & Osteopathic Physicians|Pain Medic...  
1093895625                           Dental Providers|Dentist  
1154534287                           Dental Providers|Dentist  
1902871882  Allopathic & Osteopathic Physicians|Orthopaedi...  
1003949256                Chiropractic Providers|Chiropractor  
...                                                       ...  
1164449849  Allopathic & Osteopathic Physicians|Internal M...  
1063896868  Allopathic & Osteopathic Physicians|Internal M...  
1689604191  Allopathic & Osteopathic Physicians|Internal M...  
1700808862  Allopathic & Osteopathic Physicians|Internal M...  
1760625487  Allopathic & Osteopathic Physicians|Internal M...  

[198614 rows x 8 columns]

In [ ]:
# Lets take a look at the Primary column .... I wonder what different types there are.
df_unique_doctors['Primary'].unique()

array(['Medical Doctor', 'Doctor of Dentistry', 'Chiropractor',
       'Doctor of Optometry', 'Doctor of Osteopathy',
       'Doctor of Podiatric Medicine'], dtype=object)

In [ ]:
# Primary column looks quite standard so I don't see much cleaning there.
# Lets look at the Specialty column
df_unique_doctors['Specialty'].unique()

array(['Allopathic & Osteopathic Physicians|Pain Medicine|Interventional Pain Medicine',
       'Dental Providers|Dentist',
       'Allopathic & Osteopathic Physicians|Orthopaedic Surgery',
       'Chiropractic Providers|Chiropractor',
       'Allopathic & Osteopathic Physicians|Plastic Surgery|Plastic Surgery Within the Head and Neck',
       'Allopathic & Osteopathic Physicians|Pediatrics',
       'Eye and Vision Services Providers|Optometrist',
       'Allopathic & Osteopathic Physicians|Ophthalmology',
       'Dental Providers|Dentist|Orthodontics and Dentofacial Orthopedics',
       'Allopathic & Osteopathic Physicians|Family Medicine',
       'Allopathic & Osteopathic Physicians|Pathology|Anatomic Pathology & Clinical Pathology',
       'Allopathic & Osteopathic Physicians|Pathology|Molecular Genetic Pathology',
       'Allopathic & Osteopathic Physicians|Pathology|Clinical Pathology/Laboratory Medicine',
       'Allopathic & Osteopathic Physicians|Pathology|Anatomic Pathology',


In [ ]:
# Looks like there is a main type that separates Dentists from non-Dentist Physicians, and then 1 or more specialty.
# It might make sense to remove the 'Allopathic & Osteopathic Physcians|' and 'Dental Providers|' part and just have the first of the specialties listed
specialties = df_unique_doctors.Specialty.apply(lambda x: pd.Series(str(x).split("|")))
df_unique_doctors[['GeneralSpecialty','PrimarySpecialty']]=specialties.iloc[:,0:2]

In [ ]:
df_unique_doctors.head()

Type  ProfileID FirstName  LastName  \
NPI                                                                      
1841302403  Covered Recipient Physician     557946     Vikas     Pilly   
1093895625  Covered Recipient Physician     276936   Matthew      Hall   
1154534287  Covered Recipient Physician    1275463   Stephen  Campbell   
1902871882  Covered Recipient Physician     268352     Leroy   Mccarty   
1003949256  Covered Recipient Physician     904225   Michael     Latza   

                     City State              Primary  \
NPI                                                    
1841302403  Williamsville    NY       Medical Doctor   
1093895625       Syracuse    NY  Doctor of Dentistry   
1154534287       Hinsdale    IL  Doctor of Dentistry   
1902871882          Edina    MN       Medical Doctor   
1003949256     Plainfield    NJ         Chiropractor   

                                                    Specialty  \
NPI                                                             
1841302403  Allopathic & Osteopathic Physicians|Pain Medic...   
1093895625                           Dental Providers|Dentist   
1154534287                           Dental Providers|Dentist   
1902871882  Allopathic & Osteopathic Physicians|Orthopaedi...   
1003949256                Chiropractic Providers|Chiropractor   

                               GeneralSpecialty     PrimarySpecialty  
NPI                                                                   
1841302403  Allopathic & Osteopathic Physicians        Pain Medicine  
1093895625                     Dental Providers              Dentist  
1154534287                     Dental Providers              Dentist  
1902871882  Allopathic & Osteopathic Physicians  Orthopaedic Surgery  
1003949256               Chiropractic Providers         Chiropractor

In [ ]:
df_unique_doctors.to_csv('/content/gdrive/My Drive/Team3Folder/Data/Processed/OP_2020_doctors.csv', encoding='utf-8')